### PR TITLE
[Inventory] Quantity validation on cart summary

### PR DESCRIPTION
**Pull request by Arminek**
| Q               | A
| --------------- | ---
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Related tickets | https://github.com/Sylius/Sylius/pull/5872
| License         | MIT

### CODE REVIEW

File: features/inventory/validating_product_quantity_on_checkout_complete.feature [hunk 1]
```diff
@@ -0,0 +1,35 @@
+@inventory
+Feature: Validating product quantity on checkout complete
+    In order to be sure that products I buy are available
+    As a Customer
+    I want to have them validated on checkout complete
+
+    Background:
+        Given the store operates on a single channel in "France"
```
Comment (tuka217) [inline features/inventory/validating_product_quantity_on_checkout_complete.feature]: After rebase this step no longer exists.

File: features/inventory/validating_product_quantity_on_checkout_complete.feature [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+    @ui
+    Scenario: Being unable to place an order with products that are out of stock
+        Given I have added 5 products "Iron Maiden T-Shirt" to the cart
+        And I have added 5 products "2Pac T-Shirt" to the cart
+        And I have proceeded selecting "Offline" payment method
+        When other customer has bought 7 "2Pac T-Shirt" products by this time
+        And I confirm my order
+        Then I should not see the thank you page
+        And I should be notified that this product does not have sufficient stock
```
Comment (tuka217) [inline features/inventory/validating_product_quantity_on_checkout_complete.feature]: Missing blank line

File: src/Sylius/Behat/Context/Ui/CheckoutContext.php [hunk 3]
```diff
@@ -488,6 +488,17 @@ public function iShouldSeeTheThankYouPage()
     }
 
     /**
+     * @Then I should not see the thank you page
+     */
+    public function iShouldNotSeeTheThankYouPage()
+    {
+        Assert::false(
+            $this->thankYouPage->isOpen(),
+            'I should not see thank you message, but I do'
```
Comment (tuka217) [inline src/Sylius/Behat/Context/Ui/CheckoutContext.php]: `(...) but I do*.*`

File: features/inventory/validating_product_quantity_on_checkout_complete.feature [hunk 4]
```diff
@@ -0,0 +1,43 @@
+@inventory
+Feature: Being unable to buy products that are out of stock
+    In order to be sure that products I buy are available
```
Comment (NeverResponse) [inline features/inventory/validating_product_quantity_on_checkout_complete.feature]: Either `... the product i buy...` 
or `In order to buy only available products`

File: features/inventory/validating_product_quantity_on_checkout_complete.feature [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+        And I should be notified that this product does not have sufficient stock
+
+    @ui
+    Scenario: Being unable to place an order with products that are out of stock
+        Given I have added 5 products "Iron Maiden T-Shirt" to the cart
+        And I have added 5 products "2Pac T-Shirt" to the cart
+        And I have proceeded selecting "Offline" payment method
+        When other customer has bought 7 "2Pac T-Shirt" products by this time
+        And I confirm my order
+        Then I should not see the thank you page
```
Comment (NeverResponse) [inline features/inventory/validating_product_quantity_on_checkout_complete.feature]: Not really a helpful step, besides it poses the `Where the hell am I?!` question.

File: features/inventory/validating_product_quantity_on_checkout_complete.feature (hunk 4 above)
Comment (lchrusciel) [inline features/inventory/validating_product_quantity_on_checkout_complete.feature]: `In order to ensure that products I buy are available`?

File: features/inventory/validating_product_quantity_on_checkout_complete.feature [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+        And I should be notified that this product does not have sufficient stock
+
+    @ui
+    Scenario: Being unable to place an order with products that are out of stock
+        Given I have added 5 products "Iron Maiden T-Shirt" to the cart
+        And I have added 5 products "2Pac T-Shirt" to the cart
+        And I have proceeded selecting "Offline" payment method
+        When other customer has bought 7 "2Pac T-Shirt" products by this time
```
Comment (NeverResponse) [inline features/inventory/validating_product_quantity_on_checkout_complete.feature]: How about `meanwhile another customer has bought 7 "2Pac T-Shirt" products` or
`other customer has bought 7 "2Pac T-Shirt" products in the meantime`,
whatcha think?

File: src/Sylius/Behat/Context/Transform/ProductContext.php [hunk 7]
```diff
@@ -34,9 +34,8 @@ public function __construct(ProductRepositoryInterface $productRepository)
     }
 
     /**
-     * @Transform /^product "([^"]+)"$/
-     * @Transform /^"([^"]+)" product$/
-     * @Transform /^"([^"]+)" products$/
+     * @Transform /^product(?:|s) "([^"]+)"$/
```
Comment (NeverResponse) [inline src/Sylius/Behat/Context/Transform/ProductContext.php]: `product(?:|s)` === `products?` works same looks better imo

File: features/inventory/validating_product_quantity_on_checkout_complete.feature (hunk 4 above)
Comment (NeverResponse) [inline features/inventory/validating_product_quantity_on_checkout_complete.feature]: or that ⬆️

File: src/Sylius/Behat/Context/Ui/CheckoutContext.php [hunk 8]
```diff
@@ -916,6 +927,17 @@ public function theProductShouldHaveUnitPrice(ProductInterface $product, $price)
     }
 
     /**
+     * @Given /^I should be notified that (this product) does not have sufficient stock$/
```
Comment (lchrusciel) [inline src/Sylius/Behat/Context/Ui/CheckoutContext.php]: Then

File: src/Sylius/Behat/Resources/config/suites/ui/inventory/inventory.yml [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+                - sylius.behat.context.setup.product
+                - sylius.behat.context.setup.shipping
+                - sylius.behat.context.setup.shop_security
+                - sylius.behat.context.setup.user
+
+                - sylius.behat.context.transform.addressing
+                - sylius.behat.context.transform.lexical
+                - sylius.behat.context.transform.payment
+                - sylius.behat.context.transform.product
+                - sylius.behat.context.transform.product_option
```
Comment (lchrusciel) [inline src/Sylius/Behat/Resources/config/suites/ui/inventory/inventory.yml]: Do you really use any product option in these scenarios? I couldn't find any usage.

File: src/Sylius/Behat/Resources/config/suites/ui/inventory/inventory.yml [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+        ui_inventory:
+            contexts_as_services:
+                - sylius.behat.context.hook.doctrine_orm
+
+                - sylius.behat.context.setup.channel
+                - sylius.behat.context.setup.currency
+                - sylius.behat.context.setup.geographical
+                - sylius.behat.context.setup.payment
+                - sylius.behat.context.setup.product
+                - sylius.behat.context.setup.shipping
```
Comment (lchrusciel) [inline src/Sylius/Behat/Resources/config/suites/ui/inventory/inventory.yml]: Also I didn't find any shipping context usage. Please, can you check do you really need all of these context's?

File: src/Sylius/Behat/Context/Transform/ProductContext.php (hunk 7 above)
Comment (Arminek) [inline src/Sylius/Behat/Context/Transform/ProductContext.php]: It does not work same cuz it will match both words ``products`` or ``product``.

**Comment by pjedrzejewski** [comment]
Thanks Arek! :)

**Pull request merged**